### ISSUE TITLE
Relation with `.select(...)`, `.joins(...)` and `.includes(...)` doesn't add extra selects to attributes hash

### ISSUE THREAD

**Issue by rmosolgo** [rails/rails#34889]
Hi! Thanks for reviewing my report. I'm interested in investigating a solution if you can confirm that my report reflects a real bug. 

I found some similar issues about `SELECT` with includes/joins, but it doesn't seem like those users were trying to read the value from record instances afterwards: https://github.com/rails/rails/issues/15185, https://github.com/rails/rails/issues/12423

### Steps to reproduce

- Clone https://github.com/rmosolgo/join-includes-select-bug
- `rake db:create db:migrate db:seed` 
- `rails console`

Load a relation with a custom `.select`, as well as `.joins` and `.includes`: 

```ruby
relation = Pirate.all.select("pirates.id as pirate_id").joins(:ships).includes(:ships)
  SQL (0.1ms)  SELECT  DISTINCT "pirates"."id" FROM "pirates" INNER JOIN "ships" ON "ships"."pirate_id" = "pirates"."id" LIMIT ?  [["LIMIT", 11]]
  SQL (0.4ms)  SELECT pirates.id as pirate_id, "pirates"."id" AS t0_r0, "pirates"."name" AS t0_r1, "pirates"."age" AS t0_r2, "pirates"."created_at" AS t0_r3, "pirates"."updated_at" AS t0_r4, "ships"."id" AS t1_r0, "ships"."name" AS t1_r1, "ships"."pirate_id" AS t1_r2, "ships"."created_at" AS t1_r3, "ships"."updated_at" AS t1_r4 FROM "pirates" INNER JOIN "ships" ON "ships"."pirate_id" = "pirates"."id" WHERE "pirates"."id" IN (?, ?)  [["id", 1], ["id", 2]]
```

Notice that `pirates.id as pirate_id` _was_ added to the SQL query. 

Check the attributes hash of the first record: 

```ruby
relation.first.attributes
# => {"id"=>1, "name"=>"Black Belly", "age"=>29, "created_at"=>Mon, 07 Jan 2019 14:44:08 UTC +00:00, "updated_at"=>Mon, 07 Jan 2019 14:44:08 UTC +00:00}
```

Notice it does _not_ include `pirate_id`. Accordingly, calling `.pirate_id` also fails: 

```ruby
relation.first.pirate_id
Traceback (most recent call last):
        1: from (irb):3
NoMethodError (undefined method `pirate_id' for #<Pirate:0x00007fa93b54df80>)
```

### Expected behavior

I think the attributes hash _should_ include a `"pirate_id"` key with the selected value. Accordingly, the `.pirate_id` method should return that value. 

### Actual behavior

The attributes hash has _no_ `"pirate_id"` key and calling the method raises a NoMethodError.

### System configuration
**Rails version**:

```
Rails::VERSION::STRING
# => "5.2.2"
```
**Ruby version**:

```ruby
RUBY_VERSION
# => "2.5.1"
```

Thanks again for taking a look!!

**Comment by abhaynikam** [comment]
@rmosolgo :  On the sample app for steps to reproduce throws following error on`bundle install` 

```
Authentication is required for gems.graphql.pro.
Please supply credentials for this source. You can do this by running:
 bundle config gems.graphql.pro username:password
```

Can you remove the gem `graphql-pro`?

**Comment by rmosolgo** [comment]
Derp! sorry about that. I've made a copy without that gem here: https://github.com/rmosolgo/join-includes-select-bug

And updated the instructions to reproduce the issue.

**Comment by pond** [comment]
This has been sat as a working PR with test coverage for six months. It's a useful and quite important fix. Is there a reason why it is not being merged? I'd be happy to assist the author if there are any "i"s to dot or "t"s to cross on the Rails PR submission process.

Most definitely do not want to see this stagnate! Thanks :-)

**Comment by pjrebsch** [comment]
Hi @pond, if you're referring to my PR (#35210), I never tried to message any of the maintainers about it like the contributor guide suggests because I didn't quite like how I handled parsing of the `select` alias, so I never really considered it finished by my own standards (and then I forgot about it). It does have some merge conflicts that I can take a look at to get it back to being ready for review.

**Comment by pond** [comment]
@pjrebsch It would certainly be useful, assuming the maintainers find it an acceptable approach if they do get around to looking at it. We have some location-based query stuff going on in a JSON API which pull in a "Distance" column for certain scopes, but when "includes" relationships are used this disappears. Makes the API less useful as extra calls are needed by the clients, to work around that restriction.

**Comment by cheerfulstoic** [comment]
Looks like this was released in 6.0.3! https://github.com/rails/rails/blob/v6.0.3/activerecord/CHANGELOG.md#rails-603-may-06-2020

Thanks again all!
